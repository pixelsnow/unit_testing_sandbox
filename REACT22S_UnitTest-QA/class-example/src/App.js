import './App.css';
import { useState } from 'react';

export default function App() {
  const [username, setUsername] = useState('');

  const [updated, setUpdated] = useState(username);

  const handleChange = (event) => {
    setUsername(event.target.value);
  };

  const handleClick = () => {
    console.log("button works")
    setUpdated(username);
    if (username.length < 3) {
      console.log("Username too short")
    }

    if (username.length > 20) {
      console.log("Username too long")
    }
  };
  
  return (
    <div>
      <input 
        type="text"
        id="username"
        name="username"
        onChange={handleChange}
        value={username}
      />

      <h2>Username: {username}</h2>

      <h2>Updated: {updated}</h2>

      <button onClick={handleClick}>Submit</button>
    </div>
  );
}
