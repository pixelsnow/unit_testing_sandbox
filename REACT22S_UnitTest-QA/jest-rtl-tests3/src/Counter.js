// from https://github.com/mwaz/jest-snapshot-testing
// https://circleci.com/blog/snapshot-testing-with-jest/ 

import React from "react";

class Counter extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };
    this.increment = this.increment.bind(this);
    this.decrement = this.decrement.bind(this);
  }

  increment() {
    this.setState({
      count: this.state.count + 1,
    });
  }
  decrement() {
    this.setState({
      count: Math.max(0, this.state.count - 1),
    });
  }
  render() {
    // <h1>COUNTER</h1>
    return (
      <div className="counter">
       
        <div className="buttons">
          <button onClick={this.increment}>Increment</button>
          <button onClick={this.decrement}>Decrement</button>
        </div>
        <p>{this.state.count}</p>
      </div>
    );
  }
}

export default Counter;