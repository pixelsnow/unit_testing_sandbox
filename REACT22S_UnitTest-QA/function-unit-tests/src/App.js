import './App.css';
import Calculator from './Calculator/Calculator';

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <h1>Functional Unit Tests</h1>
        <div className="Calculator">
          <h2>Calculator</h2>
          <Calculator />
        </div>
      </header>
    </div>
  );
}

export default App;
