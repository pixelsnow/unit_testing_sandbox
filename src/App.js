import { useState } from "react";

import "./App.css";

export default function App() {
  const [username, setUsername] = useState("");
  const [updated, setUpdated] = useState(username);

  const handleClick = () => {
    console.log("button works!");
    setUpdated(username);
  };

  const handleChange = (e) => {
    setUsername(e.target.value);
  };

  return (
    <div className="App">
      <input
        type="text"
        name="username"
        id="username"
        onChange={handleChange}
        value={username}
      />
      <h2>Username: {username}</h2>
      <h2>Updated: {updated}</h2>
      <button onClick={handleClick}>Submit</button>
    </div>
  );
}
