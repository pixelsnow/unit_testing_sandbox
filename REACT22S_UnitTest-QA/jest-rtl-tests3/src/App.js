// from https://github.com/mwaz/jest-snapshot-testing
// https://circleci.com/blog/snapshot-testing-with-jest/ 

import React from 'react'
import './App.css';
import Counter from './Counter'

class App extends React.Component {
  render() {

    return (
      <div className="App">
        <Counter />
      </div>
    );
  }
}

export default App;