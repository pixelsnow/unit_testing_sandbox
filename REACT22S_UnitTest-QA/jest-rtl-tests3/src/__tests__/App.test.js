import React from 'react';
import renderer from 'react-test-renderer';
// npm install --save-dev react-test-renderer
import App from '../App';
import Counter from '../Counter';


describe("Jest Snapshot testing suite", () => {
    test("Matches DOM Snapshot for App component", () => {
        const domTree = renderer.create(<App />).toJSON();
        expect(domTree).toMatchSnapshot();
    });
    // after running the first time,
    // add <h1>COUNTER</h1> to Counter.js render()
    // run tests again -> snapshot broken
    // use "u" option in Jest to update the shapshot

    test("Should render 3 after three increments", () => {
        const component = renderer.create(<Counter />);
        component.getInstance().increment();
        component.getInstance().increment();
        component.getInstance().increment();
        expect(component.toJSON()).toMatchSnapshot();
    });

});

const sum = require('../math/sum');
describe("Testing multiple cases", () => {
    const testValues = [
        //[numA,numB,result]
        [5,6,11],
        [5,-6,-1],
        [-5,6,1],
        [-5,-6,-11],
        [0,0,0]
    ];
    
    // test.each() uses printf formatting (example: %s)
    // https://jestjs.io/docs/api#testeachtablename-fn-timeout
    test.each(testValues)("add %s and %s to equal %s", (numA,numB,result)=> {
        expect(sum(numA,numB)).toBe(result);
    });

    test('adds 1 + 2 to equal 3', () => {
        expect(sum(1,2)).toBe(3)
    });
});

describe("Testing error messages", () => {
    const testValuesA = [
        //[numA,errorA]
        ["","Error: sum() value A is not a number"],
        [,"Error: sum() value A is not a number"],
        ["h","Error: sum() value A is not a number"],
        ["2","Error: sum() value A is not a number"],
        ["*","Error: sum() value A is not a number"]
    ];

    test.each(testValuesA)('sum() value A "%s" throws the exception "%s"', (numA,errorA) => {
        expect(()=>new sum(numA,2)).toThrow(errorA);
        //wrapping the sum() in an assertion makes it so we can actually catch the error
        //https://jestjs.io/docs/expect#tothrowerror
    });
});