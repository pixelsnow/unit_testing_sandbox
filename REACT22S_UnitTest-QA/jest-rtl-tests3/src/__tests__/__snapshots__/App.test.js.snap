// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Snapshot testing suite Matches DOM Snapshot for App component 1`] = `
<div
  className="App"
>
  <div
    className="counter"
  >
    <h1>
      COUNTER
    </h1>
    <div
      className="buttons"
    >
      <button
        onClick={[Function]}
      >
        Increment
      </button>
      <button
        onClick={[Function]}
      >
        Decrement
      </button>
    </div>
    <p>
      0
    </p>
  </div>
</div>
`;

exports[`Jest Snapshot testing suite Should render 3 after three increments 1`] = `
<div
  className="counter"
>
  <h1>
    COUNTER
  </h1>
  <div
    className="buttons"
  >
    <button
      onClick={[Function]}
    >
      Increment
    </button>
    <button
      onClick={[Function]}
    >
      Decrement
    </button>
  </div>
  <p>
    3
  </p>
</div>
`;

exports[`Jest Snapshot testing suite Should render 3 after three increments 2`] = `
<div
  className="counter"
>
  <h1>
    COUNTER
  </h1>
  <div
    className="buttons"
  >
    <button
      onClick={[Function]}
    >
      Increment
    </button>
    <button
      onClick={[Function]}
    >
      Decrement
    </button>
  </div>
  <p>
    3
  </p>
</div>
`;
